app/src/main/kotlin/Twenty48/Game2048.kt

```diff
@@ -0,0 +1,177 @@
+            arrayOf(0, 0, 0, 0),
+            arrayOf(0, 0, 0, 0),
+            arrayOf(0, 0, 0, 0),
+            arrayOf(0, 0, 0, 0)
+        )
+
+        val gameOverMessage = run2048(grid)
+        println(gameOverMessage)
+    }
+
+    fun run2048(grid: Array<Array<Int>>): String {
+        if (isGridSolved(grid)) return positiveGameOverMessage
+        else if (isGridFull(grid)) return negativeGameOverMessage
+
+        val populatedGrid = spawnNumber(grid)
+        display(populatedGrid)
+
+        return run2048(manipulateGrid(populatedGrid, waitForValidInput()))
+    }
+
+    fun isGridSolved(grid: Array<Array<Int>>): Boolean = grid.any { row -> row.contains(2048) }
+    fun isGridFull(grid: Array<Array<Int>>): Boolean = grid.all { row -> !row.contains(0) }
+
+    fun spawnNumber(grid: Array<Array<Int>>): Array<Array<Int>> {
+        val coordinates = locateSpawnCoordinates(grid)
+        val number = generateNumber()
+
+        return updateGrid(grid, coordinates, number)
+    }
+
+    fun locateSpawnCoordinates(grid: Array<Array<Int>>): Pair<Int, Int> {
+        val emptyCells = arrayListOf<Pair<Int, Int>>()
+        grid.forEachIndexed { x, row ->
+            row.forEachIndexed { y, cell ->
+                if (cell == 0) emptyCells.add(Pair(x, y))
+            }
+        }
+
+        return emptyCells[(Math.random() * (emptyCells.size - 1)).toInt()]
+    }
+
+    fun generateNumber(): Int = if (Math.random() > 0.10) 2 else 4
+
+    fun updateGrid(grid: Array<Array<Int>>, at: Pair<Int, Int>, value: Int): Array<Array<Int>> {
+        val updatedGrid = grid.copyOf()
+        updatedGrid[at.first][at.second] = value
+        return updatedGrid
+    }
+
+    fun waitForValidInput(): String {
+        val input = waitForInput()
+        return if (isValidInput(input)) input else waitForValidInput()
+    }
+
+    fun isValidInput(input: String): Boolean = arrayOf("up", "left", "right", "down").contains(input)
+
+    fun waitForInput(): String {
+        val reader = BufferedReader(InputStreamReader(System.`in`))
+        println("Direction?  ")
+        return reader.readLine()
+    }
+
+    fun manipulateGrid(grid: Array<Array<Int>>, input: String): Array<Array<Int>> = when (input) {
+        "left" -> shiftCellsLeft(grid)
+        "down" -> shiftCellsDown(grid)
+        "right" -> shiftCellsRight(grid)
+        "up" -> shiftCellsUp(grid)
+        else -> throw IllegalArgumentException("Expected one of [up, left, right, down]")
+    }
+
+    fun shiftCellsLeft(grid: Array<Array<Int>>): Array<Array<Int>> =
+        grid.map(::mergeAndOrganizeCells).toTypedArray()
+
+    fun shiftCellsRight(grid: Array<Array<Int>>): Array<Array<Int>> =
+        grid.map { row -> mergeAndOrganizeCells(row.reversed().toTypedArray()).reversed().toTypedArray() }.toTypedArray()
+
+    fun shiftCellsUp(grid: Array<Array<Int>>): Array<Array<Int>> {
+        val rows: Array<Array<Int>> = arrayOf(
+            arrayOf(grid[0][0], grid[1][0], grid[2][0], grid[3][0]),
```

Comment:
I'm not a fan of hardcoding these. What if I wanted to make a 5x5 grid game. Algos shouldn't be fixed to a specific parameter that's not passed in or is a member variable.

Reply:
This implementation is from [Rosetta Code](https://rosettacode.org/wiki/2048) actually. They hardcoded in the Grid, and it could require an entire rewrite to refactor. The Scala and Java implementations seem modular, we could use those too I guess.